### PR TITLE
enhance sdl2 pixelformat selection

### LINKED ISSUE #202: sdl2 does not convert pixel formats properly?
Dear Alfred,

I've just tried to play around a little with nvenc/cuvid & avformat and stumbled across another issue/lack with sdl2. When using cuvid to decode the pixel format seems to be wrong. The picture looks similar to what I've seen during implementing nv12_to_rgb32.

Attached please find a screenshot of the sdl2 displayed picture when decoded with h264_cuvid.
baresip_sdl2_cuvid

Setup is as follows: baresip (avformat,file & avcodec_h264enc h264_nvenc) <-> baresip (avcodec_h264dec h264_cuvid / sdl2)

### PR DESCRIPTION
**Pull request by hargut**
fixes #202 and #201

### CODE REVIEW

File: modules/sdl2/sdl.c [hunk 1]
```diff
@@ -30,6 +30,25 @@ struct vidisp_st {
...
+	case VID_FMT_UYVY422:	return SDL_PIXELFORMAT_UYVY;
+	case VID_FMT_RGB32:	return SDL_PIXELFORMAT_RGBA8888;
+	case VID_FMT_ARGB:	return SDL_PIXELFORMAT_ARGB8888;
+	case VID_FMT_RGB565:	return SDL_PIXELFORMAT_RGB565;
+	case VID_FMT_RGB555:	return SDL_PIXELFORMAT_RGB555;
+	case VID_FMT_NV12:	return SDL_PIXELFORMAT_NV12;
+	case VID_FMT_NV21:	return SDL_PIXELFORMAT_NV21;
+	case VID_FMT_YUV444P:
+	case VID_FMT_N:
+	default:		return SDL_PIXELFORMAT_UNKNOWN;
```
Comment (alfredh) [inline modules/sdl2/sdl.c]: have you tested that all these pixel formats are working?

we dont have to add all formats now, I think it is better to just add a couple of
formats that is tested and working, such as YUV420P and NV12.

File: modules/sdl2/sdl.c [hunk 2]
```diff
@@ -30,6 +30,25 @@ struct vidisp_st {
 static struct vidisp *vid;
 
 
+static enum vidfmt match_fmt(u_int32_t fmt)
```
Comment (alfredh) [inline modules/sdl2/sdl.c]: the input parameter type should be: `enum vidfmt`

the return type should be `Uint32` (same as input to `SDL_CreateTexture`).

**Comment by alfredh** [comment]
this is a nice improvement. I think the long list of pixel formats can be shorter.

we also need to update the code that copies pixels from vidframe into the SDL texture:

https://github.com/alfredh/baresip/blob/master/modules/sdl2/sdl.c#L174

the code assumes YUV420P. it needs to handle that `data[i]` and `linesize[i]`
is zero for some formats.

for example when I tested your patch locally, an input frame with NV12
causes a crash in the pixel copying code.

**Comment by hargut** [comment]
Dear Alfred,

to be honest, in this shape this shouldn't have appeared as pull request. The patch wasn't even working.

I've tried to fix the issue in a rush, and have overlooked that h264_decoder was not set properly to h264_cuvid. Therefore I thought everything was working during the test, but it simply wasn't as there where no NV12 frames input to the sdl2 module.

Thank you for your efforts to fix it!

**Pull request closed without merging**